ProduceContentAssets should not "url decode" paths provided by NuGet

Create the following project and run `dotnet build`.
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net462</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="ContentFilesExample" Version="*" />
  </ItemGroup>

</Project>
```

Make sure that nuget.org is among the sources. 

```
 dotnet build
Microsoft (R) Build Engine version 16.0.450+ga8dc7f1d34 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Restore completed in 530.45 ms for C:\Users\nikolev.REDMOND\Source\Repos\ContentFilesExampleTest\ContentFilesExampleTest\ContentFilesExampleTest.csproj.
C:\Program Files\dotnet\sdk\2.2.202\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(350,5): error MSB4018: The "ProduceContentAssets" task failed unexpectedly. [C:\Users\nikolev.REDMOND\Source\Repos\ContentFilesExampleTest\ContentFilesExampleTest\ContentFilesExampleTest.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(350,5): error MSB4018: System.IO.DirectoryNotFoundException: Could not find a part of the path 'C:\Users\nikolev.REDMOND\Source\Repos\ContentFilesExampleTest\globalPackagesFolder\contentfilesexample\1.0.2\contentFiles\cs\portable-net45+wpa81+wp8+win8\ExampleReader.cs.pp'. [C:\Users\nikolev.REDMOND\Source\Repos\ContentFilesExampleTest\ContentFilesExampleTest\ContentFilesExampleTest.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(350,5): error MSB4018:    at System.IO.FileStream.ValidateFileHandle(SafeFileHandle fileHandle) [C:\Users\nikolev.REDMOND\Source\Repos\ContentFilesExampleTest\ContentFilesExampleTest\ContentFilesExampleTest.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(350,5): error MSB4018:    at System.IO.FileStream.CreateFileOpenHandle(FileMode mode, FileShare share, FileOptions options) [C:\Users\nikolev.REDMOND\Source\Repos\ContentFilesExampleTest\ContentFilesExampleTest\ContentFilesExampleTest.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(350,5): error MSB4018:    at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options) [C:\Users\nikolev.REDMOND\Source\Repos\ContentFilesExampleTest\ContentFilesExampleTest\ContentFilesExampleTest.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(350,5): error MSB4018:    at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share) [C:\Users\nikolev.REDMOND\Source\Repos\ContentFilesExampleTest\ContentFilesExampleTest\ContentFilesExampleTest.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(350,5): error MSB4018:    at System.IO.File.OpenRead(String path) [C:\Users\nikolev.REDMOND\Source\Repos\ContentFilesExampleTest\ContentFilesExampleTest\ContentFilesExampleTest.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(350,5): error MSB4018:    at Microsoft.NET.Build.Tasks.NugetContentAssetPreprocessor.Process(String originalAssetPath, String relativeOutputPath, String& pathToFinalAsset) [C:\Users\nikolev.REDMOND\Source\Repos\ContentFilesExampleTest\ContentFilesExampleTest\ContentFilesExampleTest.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(350,5): error MSB4018:    at Microsoft.NET.Build.Tasks.ProduceContentAssets.ProduceContentAsset(ITaskItem contentFile) [C:\Users\nikolev.REDMOND\Source\Repos\ContentFilesExampleTest\ContentFilesExampleTest\ContentFilesExampleTest.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(350,5): error MSB4018:    at Microsoft.NET.Build.Tasks.ProduceContentAssets.ExecuteCore() [C:\Users\nikolev.REDMOND\Source\Repos\ContentFilesExampleTest\ContentFilesExampleTest\ContentFilesExampleTest.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(350,5): error MSB4018:    at Microsoft.NET.Build.Tasks.TaskBase.Execute() [C:\Users\nikolev.REDMOND\Source\Repos\ContentFilesExampleTest\ContentFilesExampleTest\ContentFilesExampleTest.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(350,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() [C:\Users\nikolev.REDMOND\Source\Repos\ContentFilesExampleTest\ContentFilesExampleTest\ContentFilesExampleTest.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(350,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask) [C:\Users\nikolev.REDMOND\Source\Repos\ContentFilesExampleTest\ContentFilesExampleTest\ContentFilesExampleTest.csproj]

Build FAILED.

C:\Program Files\dotnet\sdk\2.2.202\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(350,5): error MSB4018: The "ProduceContentAssets" task failed unexpectedly. [C:\Users\nikolev.REDMOND\Source\Repos\ContentFilesExampleTest\ContentFilesExampleTest\ContentFilesExampleTest.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(350,5): error MSB4018: System.IO.DirectoryNotFoundException: Could not find a part of the path 'C:\Users\nikolev.REDMOND\Source\Repos\ContentFilesExampleTest\globalPackagesFolder\contentfilesexample\1.0.2\contentFiles\cs\portable-net45+wpa81+wp8+win8\ExampleReader.cs.pp'. [C:\Users\nikolev.REDMOND\Source\Repos\ContentFilesExampleTest\ContentFilesExampleTest\ContentFilesExampleTest.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(350,5): error MSB4018:    at System.IO.FileStream.ValidateFileHandle(SafeFileHandle fileHandle) [C:\Users\nikolev.REDMOND\Source\Repos\ContentFilesExampleTest\ContentFilesExampleTest\ContentFilesExampleTest.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(350,5): error MSB4018:    at System.IO.FileStream.CreateFileOpenHandle(FileMode mode, FileShare share, FileOptions options) [C:\Users\nikolev.REDMOND\Source\Repos\ContentFilesExampleTest\ContentFilesExampleTest\ContentFilesExampleTest.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(350,5): error MSB4018:    at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options) [C:\Users\nikolev.REDMOND\Source\Repos\ContentFilesExampleTest\ContentFilesExampleTest\ContentFilesExampleTest.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(350,5): error MSB4018:    at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share) [C:\Users\nikolev.REDMOND\Source\Repos\ContentFilesExampleTest\ContentFilesExampleTest\ContentFilesExampleTest.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(350,5): error MSB4018:    at System.IO.File.OpenRead(String path) [C:\Users\nikolev.REDMOND\Source\Repos\ContentFilesExampleTest\ContentFilesExampleTest\ContentFilesExampleTest.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(350,5): error MSB4018:    at Microsoft.NET.Build.Tasks.NugetContentAssetPreprocessor.Process(String originalAssetPath, String relativeOutputPath, String& pathToFinalAsset) [C:\Users\nikolev.REDMOND\Source\Repos\ContentFilesExampleTest\ContentFilesExampleTest\ContentFilesExampleTest.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(350,5): error MSB4018:    at Microsoft.NET.Build.Tasks.ProduceContentAssets.ProduceContentAsset(ITaskItem contentFile) [C:\Users\nikolev.REDMOND\Source\Repos\ContentFilesExampleTest\ContentFilesExampleTest\ContentFilesExampleTest.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(350,5): error MSB4018:    at Microsoft.NET.Build.Tasks.ProduceContentAssets.ExecuteCore() [C:\Users\nikolev.REDMOND\Source\Repos\ContentFilesExampleTest\ContentFilesExampleTest\ContentFilesExampleTest.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(350,5): error MSB4018:    at Microsoft.NET.Build.Tasks.TaskBase.Execute() [C:\Users\nikolev.REDMOND\Source\Repos\ContentFilesExampleTest\ContentFilesExampleTest\ContentFilesExampleTest.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(350,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() [C:\Users\nikolev.REDMOND\Source\Repos\ContentFilesExampleTest\ContentFilesExampleTest\ContentFilesExampleTest.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(350,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask) [C:\Users\nikolev.REDMOND\Source\Repos\ContentFilesExampleTest\ContentFilesExampleTest\ContentFilesExampleTest.csproj]
    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:01.67
```

The issue here is that the selected framework for the contentfiles is a PCL framework and `+` in the name, encoded as `%2B`. 

NuGet extracts the encoded name, and writes that to the assets file, g.props as needed.

Paths that the targets read from NuGet should be treated as strings and not decoded in any way.

--- a/eng/Versions.props
+++ b/eng/Versions.props
@@ -11,6 +11,7 @@
   </PropertyGroup>
   <!-- Production Dependencies -->
   <PropertyGroup>
+    <MicrosoftBuildVersion>15.4.8</MicrosoftBuildVersion>
     <MicrosoftBuildFrameworkVersion>15.4.8</MicrosoftBuildFrameworkVersion>
     <MicrosoftBuildUtilitiesCoreVersion>15.4.8</MicrosoftBuildUtilitiesCoreVersion>
     <MicrosoftExtensionsDependencyModelVersion>2.1.0-preview2-26306-03</MicrosoftExtensionsDependencyModelVersion>

--- a/src/Tasks/Microsoft.NET.Build.Tasks/Microsoft.NET.Build.Tasks.csproj
+++ b/src/Tasks/Microsoft.NET.Build.Tasks/Microsoft.NET.Build.Tasks.csproj
@@ -41,6 +41,7 @@
   </ItemDefinitionGroup>
 
   <ItemGroup>
+    <PackageReference Include="Microsoft.Build" Version="$(MicrosoftBuildVersion)" ExcludeAssets="Runtime" />
     <PackageReference Include="Microsoft.Build.Framework" Version="$(MicrosoftBuildFrameworkVersion)" ExcludeAssets="Runtime" />
     <PackageReference Include="Microsoft.Build.Utilities.Core" Version="$(MicrosoftBuildUtilitiesCoreVersion)" ExcludeAssets="Runtime" />
     <PackageReference Include="Microsoft.Extensions.DependencyModel" Version="$(MicrosoftExtensionsDependencyModelVersion)" />

--- a/src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs
+++ b/src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs
@@ -8,6 +8,7 @@ using System.IO;
 using System.Linq;
 using System.Security.Cryptography;
 using System.Text;
+using Microsoft.Build.Evaluation;
 using Microsoft.Build.Framework;
 using Microsoft.Build.Utilities;
 using NuGet.Common;
@@ -1129,8 +1130,7 @@ namespace Microsoft.NET.Build.Tasks
             {
                 FlushMetadata();
                 _itemCount++;
-                _writer.Write(itemSpec);
-
+                _writer.Write(ProjectCollection.Escape(itemSpec));
             }
 
             private void WriteItem(string itemSpec, LockFileTargetLibrary package)

--- a/src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildANetCoreApp.cs
+++ b/src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildANetCoreApp.cs
@@ -673,6 +673,57 @@ public static class Program
                 .Pass();
         }
 
+        [WindowsOnlyFact]
+        public void It_escapes_resolved_package_assets_paths()
+        {
+            var testProject = new TestProject()
+            {
+                Name = "ProjectWithPackageThatNeedsEscapes",
+                TargetFrameworks = "net462",
+                IsSdkProject = true,
+                IsExe = true,
+            };
+
+            testProject.SourceFiles["ExampleReader.cs"] = @"
+using System;
+using System.Threading.Tasks;
+
+namespace ContentFilesExample
+{
+    internal static class ExampleInternals
+    {
+        internal static Task<string> GetFileText(string fileName)
+        {
+            throw new NotImplementedException();
+        }
+    }
+}
+
+class Program
+{
+    static void Main(string[] args)
+    {
+        Console.WriteLine(""Hello World!"");
+    }
+}";
+
+            // ContentFilesExample is an existing package that demonstrates the problem.
+            // It contains assets with paths that have '%2B', which MSBuild will unescape to '+'.
+            // Without the change to escape the asset paths, the asset will not be found inside the package.
+            testProject.PackageReferences.Add(new TestPackageReference("ContentFilesExample", "1.0.2"));
+
+            var testAsset = _testAssetsManager
+                .CreateTestProject(testProject)
+                .Restore(Log, testProject.Name);
+
+            var buildCommand = new BuildCommand(Log, Path.Combine(testAsset.TestRoot, testProject.Name));
+
+            buildCommand
+                .Execute()
+                .Should()
+                .Pass();
+        }
+
         private static bool IsPE32(string path)
         {
             using (var reader = new PEReader(File.OpenRead(path)))

